daemon/health.go

```diff
@@ -248,12 +248,31 @@ func handleProbeResult(d *Daemon, c *container.Container, result *types.Healthch
 // There is never more than one monitor thread running per container at a time.
 func monitor(d *Daemon, c *container.Container, stop chan struct{}, probe probe) {
 	probeInterval := timeoutWithDefault(c.Config.Healthcheck.Interval, defaultProbeInterval)
+	startInterval := timeoutWithDefault(c.Config.Healthcheck.StartInterval, defaultProbeInterval)
+	startPeriod := timeoutWithDefault(c.Config.Healthcheck.StartPeriod, defaultStartPeriod)
 
 	intervalTimer := time.NewTimer(probeInterval)
 	defer intervalTimer.Stop()
 
+	c.Lock()
+	started := c.State.StartedAt
+	var status string
+	if c.State.Health != nil {
+		status = c.State.Health.Health.Status
+	}
+	c.Unlock()
+
 	for {
-		intervalTimer.Reset(probeInterval)
+		if time.Now().Sub(started) < startPeriod && status == types.Starting {
+			c.Lock()
+			if c.State.Health != nil {
+				status = c.State.Health.Health.Status
+			}
+			c.Unlock()
+			intervalTimer.Reset(startInterval)
```

Comment:
The value of `status` on line 266 is the status from the previous iteration of the loop, before the most recent health probe. Therefore if the container becomes healthy while still within the start period, there will be one additional probe at `startInterval` after the container transitions to healthy. The decision on which interval to use before the next probe should be based on the most recent probe result, not a stale one.

```go
c.Lock()
started := c.State.StartedAt
status := c.State.Health.Status()
c.Unlock()

var intervalTimer *time.Timer
if time.Now().Sub(started) < startPeriod && status == types.Starting {
	intervalTimer = time.NewTimer(startInterval)
} else {
	intervalTimer = time.NewTimer(probeInterval)
}

for {
	select {
	/* ... */
	case <-intervalTimer.C:
		/* ... */
		select {
		case <-stop:
			/* ... */
		case result := <-results:
			status = handleProbeResult(d, c, result, stop) // Modify handleProbeResult to return the `current` status
			cancelProbe()
		}
	}
	if status == types.Starting && time.Now().Sub(started) < startPeriod {
		intervalTimer.Reset(startInterval)
	} else {
		intervalTimer.Reset(probeInterval)
	}
}
```

Reply:
I've updated this let me know what you think.
I moved the interval selection behind a function that, I think, makes it a bit easier to read/reason about.